Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -818,7 +818,32 @@ void RenderBlockFlow::layoutBlockChildren(bool relayoutChildren, LayoutUnit& max
         // Lay out the child.
         layoutBlockChild(child, marginInfo, previousFloatLogicalBottom, maxFloatLogicalBottom);
     }
-    
+
+    // If we are trimming the block end margin, we need to make sure we actually trim the margin of the children
+    // at the end of the block by walking back up the container. Any self collapsing children will also need to
+    // have their position adjusted to below the last non self-collapsing child.
```

Comment:
Are you planning on transitioning this to handle descendant block boxes with collapsed through margins? Within a BFC, you can have arbitrary number of block boxes at arbitrary nesting levels.

Reply:
Yes you are right. I think I will try to address that in a separate patch since I can see that being slightly more complicated (unless you think both scenarios should be addressed in a single patch). I'll make the necessary changes to the patch/bug report information

Reply:
While this patch fixes these test included cases, real web content tend to be slightly more complicated with block content nesting. However I am ok with landing this patch as is, if we need to quickly address a broken popular content.  Now if this is just regular fix, I'd try to figure out first how nested cases would be handled as they may require a very different approach e.g. instead of back-tracing you may be building a list of block boxes that touch the block-end side of the BFC root as block layout progresses -as by definition the in-flow block level box that is being laid out is adjacent to the BFC root's content box (bottom) (under certain conditions).

Reply:
Modified the patch to try and detect this type of nesting and recurse into the nested block container and perform the same logic. I think the number of scenarios in which type of nesting can occur and have an impact on the block-end margins is more limited than I initially thought there were going to be so I provided my reasoning in the writeup with an example, but correct me if I'm wrong!

Reply:
> I think the number of scenarios in which type of nesting can occur and have an impact on the block-end margins is more limited than I initially thought

You surely thought about it a lot more than I did, so I am curious of what scenarios you had in your mind (that turned out to be non-issues). Anyway when I made that comment I was just thinking along the lines of arbitrary number of block level boxes with collapsed through margins at arbitrary level of nesting....which may make you consider a different approach to ensure performant margin-trimming (back-tracing vs. tracking boxes with collapsed through margins).
It looks like you decided to continue with back-tracing.
> This means we need to walk back up the children of the container and perform these  adjustments until we hit a child that is not self-collapsing

Is that because you think authors don't create such arbitrarily expensive content or some other reasons? (I personally don't think they do, but I also don't have any data to back up such claims).  
If we stick with this approach I think the preferred way of doing it is a non-recursive style.